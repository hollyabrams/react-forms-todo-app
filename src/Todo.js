import { useState } from 'react';
import './Todo.scss';

function Todo({
  id,
  task,
  completed,
  removeTodo,
  editTodo,
  toggleCompletion,
}) {
  const [editing, setEditing] = useState(false);
  const [newTask, setNewTask] = useState(task);

  const handleRemove = () => {
    removeTodo(id);
  };

  const handleEdit = () => {
    setEditing(true);
  };

  const handleUpdate = (e) => {
    e.preventDefault();
    editTodo(id, newTask);
    setEditing(false);
  };

  const handleToggleCompletion = () => {
    toggleCompletion(id);
  };

  const todoStyle = {
    textDecoration: completed ? 'line-through' : 'none',
    cursor: 'pointer',
  };

  return (
    <li className={`todo ${completed ? 'completed' : ''}`}>
      {editing ? (
        <form onSubmit={handleUpdate} className="d-flex align-items-center">
          <input
            type="text"
            value={newTask}
            onChange={(e) => setNewTask(e.target.value)}
            className="form-control mr-3"
          />
          <button type="submit" className="todo-item__buttons__toggle-completion-button">
            Save
          </button>
        </form>
      ) : (
        <div className="d-flex justify-content-between align-items-center">
          <span
            style={todoStyle}
            onClick={handleToggleCompletion}
            className="flex-grow-1 mr-3"
          >
            {task}
          </span>
          <div className="d-flex">
            <button
              onClick={handleEdit}
              className="todo-item__buttons__edit-button"
            >
              Edit
            </button>
            <button onClick={handleRemove} className="todo-item__buttons__remove-button">
              X
            </button>
            <button
              onClick={handleToggleCompletion}
              className="todo-item__buttons__toggle-completion-button"
            >
              {completed ? 'Remove Strike-Through' : 'Mark as Completed'}
            </button>
          </div>
        </div>
      )}
    </li>
  );
}

export default Todo;



