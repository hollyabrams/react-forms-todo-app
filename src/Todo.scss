.todo {
    margin: 10px;
    padding: 10px;
    font-size: 1rem;
    border: 1px solid #ccc;
    border-radius: 4px;
    margin-bottom: 10px;
    display: flex;
    justify-content: space-between;
    align-items: center;
    background-color: #fff;
    box-shadow: 0px 3px 3px #ccc;
    transition: all 0.3s ease;

    &:hover {
        box-shadow: 0px 3px 5px #ccc;
    }

    & span {
        flex-grow: 1;
        padding: 10px;
        cursor: pointer;
        text-decoration: none;
        color: #333;

        &.completed {
            text-decoration: line-through;
            color: #ccc;
        }
    }

    & button {
        margin: 10px;
    }

    & .edit-form {
        display: flex;
        align-items: center;
        margin: 10px;
        padding: 10px;
        border-radius: 4px;
        background-color: #f0f0f0;

        & input[type='text'] {
            flex-grow: 1;
            padding: 5px;
            margin-right: 10px;
            border-radius: 4px;
            border: none;
            background-color: #e0e0e0;

            &:focus {
                outline: none;
                background-color: #fff;
                box-shadow: 0px 0px 3px #ccc;
            }
        }

        & button[type='submit'] {
            background-color: #28a745;
            border: none;
            color: #fff;
            border-radius: 4px;
            padding: 5px 15px;

            &:hover {
                background-color: #218838;
            }
        }

        & button[type='button'] {
            background-color: #dc3545;
            border: none;
            color: #fff;
            border-radius: 4px;
            padding: 5px 15px;
            margin-left: 10px;

            &:hover {
                background-color: #c82333;
            }
        }
    }
}