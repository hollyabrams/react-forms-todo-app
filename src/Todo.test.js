import { render } from '@testing-library/react';
import Todo from './Todo';

describe('Todo', () => {
  const todo = {
    id: 1,
    task: 'Test Todo',
    completed: false,
    removeTodo: jest.fn(),
    editTodo: jest.fn(),
    toggleCompletion: jest.fn(),
  };

  it('should render without crashing', () => {
    render(<Todo {...todo} />);
  });

  it('should match snapshot', () => {
    const { container } = render(<Todo {...todo} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
