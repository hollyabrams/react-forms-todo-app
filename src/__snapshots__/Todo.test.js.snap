// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo should match snapshot 1`] = `
<li
  class="todo "
>
  <div
    class="d-flex justify-content-between align-items-center"
  >
    <span
      class="flex-grow-1 mr-3"
      style="text-decoration: none; cursor: pointer;"
    >
      Test Todo
    </span>
    <div
      class="d-flex"
    >
      <button
        class="todo-item__buttons__edit-button"
      >
        Edit
      </button>
      <button
        class="todo-item__buttons__remove-button"
      >
        X
      </button>
      <button
        class="todo-item__buttons__toggle-completion-button"
      >
        Mark as Completed
      </button>
    </div>
  </div>
</li>
`;
